Update README (#103)
buildkite/test-splitter: README.md

# Buildkite Test Splitter

> Buildkite Test Splitter is an open source tool to orchestrate your test suites. It uses your Buildkite Test Analytic suite data to intelligently partition and parallelise your tests. 

Reviewer: ```suggestion
Buildkite Test Splitter is an open source tool to orchestrate your test suites. It uses your Buildkite Test Analytics suite data to intelligently partition and parallelise your tests. 
```

## Using the Test Splitter

### ENV variables
Test Splitter uses the following environment variables provided in your Buildkite build environment. 
Please make sure that the following environment variables are available in your testing environment, particularly if you use Docker or some other type of containerization to run your tests.
| Environment Variable | Description|
| -------------------- | ----------- |
| `BUILDKITE_BUILD_ID` | The UUID of the Buildkite build. Test Splitter uses this UUID along with `BUILDKITE_STEP_ID` to uniquely identify the test plan. |
| `BUILDKITE_JOB_ID` | The UUID of the job in Buildkite build. |
| `BUILDKITE_ORGANIZATION_SLUG` | The slug of your Buildkite organization. |

> | `BUILDKITE_PARALLEL_JOB` | The index of parallel job created from a Buildkite parallel build step. Test Splitter uses this value to get specific test plan for each parallel job. <br>Make sure you configure `parallelism` for your Buildkite build step.  You can read more on Buildkite parallel build step on this [page](https://buildkite.com/docs/pipelines/controlling-concurrency#concurrency-and-parallelism).| 

Reviewer: ```suggestion
| `BUILDKITE_PARALLEL_JOB` | The index number of a parallel job created from a Buildkite parallel build step. Make sure you configure `parallelism` in your pipeline definition.  You can read more about parallel build steps on this [page](https://buildkite.com/docs/pipelines/controlling-concurrency#concurrency-and-parallelism).| 
```

> | `BUILDKITE_PARALLEL_JOB_COUNT` | The total number of parallel jobs created from a Buildkite parallel build step. Test Splitter uses this value to split the tests and create the test plan. <br>Make sure you configure `parallelism` for your Buildkite build step.  You can read more on Buildkite parallel build step on this [page](https://buildkite.com/docs/pipelines/controlling-concurrency#concurrency-and-parallelism) |

Reviewer: ```suggestion
| `BUILDKITE_PARALLEL_JOB_COUNT` | The total number of parallel jobs created from a Buildkite parallel build step. Make sure you configure `parallelism` in your pipeline definition.  You can read more about Buildkite parallel build step on this [page](https://buildkite.com/docs/pipelines/controlling-concurrency#concurrency-and-parallelism) |
```

> | `BUILDKITE_STEP_ID` | The UUID of the step group in Buildkite build. Test Splitter uses this UUID along with `BUILDKITE_BUILD_ID` to uniquely identify the test plan.

Reviewer: I understand that you've ordered this list alphabetically, but I think it might make more sense to group them semantically? i.e. 
BUILDKITE_BUILD_ID
BUILDKITE_STEP_ID
BUILDKITE_ORGANIZATION_SLUG
BUILDKITE_PARALLEL_JOB_COUNT
BUILDKITE_PARALLEL_JOB

And then you won't have to repeat the explanations all the time?

Author: I'll keep it sorted alphabetically to be consistent with other table and a practice we do in our [docs](https://buildkite.com/docs/pipelines/environment-variables).

We will have to repeat the explanations anyway since it is a table format. We will have to change the format, or merge some rows to have a single description.

<br>

> In addition to above variables, you must set following environment variables.

Reviewer: I would put this section first, before the ones that are generated by Buildkite

Author: I'll keep it this way. I think it more sense to explain the basis first, which is the ENV provided by Buildkite. This is also consistent with the instruction in [setup page](https://docs.google.com/document/d/1aDJovqDU1KPvlmIETjPN2G3Pvn4xjo9ZD8vVK2RbGog/edit).

| Environment Variable | Description |
| -------------------- | ----------- |
| `BUILDKITE_SPLITTER_API_ACCESS_TOKEN ` | Buildkite API access token with `read_suites`, `read_test_plan`, and `write_test_plan` scopes. You can create access token from [Personal Settings](https://buildkite.com/user/api-access-tokens) in Buildkite |

> | `BUILDKITE_SPLITTER_SUITE_SLUG` | The slug of your Buildkite Test Analytics test suite. You can find the suite slug in the url for your suite, for example, the slug for the url: https://buildkite.com/organizations/my-organization/analytics/suites/my-suite is `my-suite` |

Reviewer: ```suggestion
| `BUILDKITE_SPLITTER_SUITE_SLUG` | The slug of your Buildkite Test Analytics test suite. You can find the suite slug in the url for your suite. For example, the slug for the url: https://buildkite.com/organizations/my-organization/analytics/suites/my-suite is `my-suite` |
```

<br>

> Following environment variables can be used optionally to configure your test splitter.

Reviewer: ```suggestion
The following environment variables can be used optionally to configure your test splitter.
```

| Environment Variable | Default Value | Description |
| ---- | ---- | ----------- |
| `BUILDKITE_SPLITTER_DEBUG_ENABLED` | `false` | Flag to enable more verbose logging. |

> | `BUILDKITE_SPLITTER_RETRY_COUNT` | `0` | The number of retry. Test splitter runs the test command defined in `BUILDKITE_SPLITTER_TEST_CMD`, and retries the failing tests maximum `BUILDKITE_SPLITTER_RETRY_COUNT` times. For Rspec, the test splitter runs `BUILDKITE_SPLITTER_TEST_CMD` with `--only-failures` as the retry command. |

Reviewer: ```suggestion
| `BUILDKITE_SPLITTER_RETRY_COUNT` | `0` | The number of retries permitted. Test splitter runs the test command defined in `BUILDKITE_SPLITTER_TEST_CMD`, and retries only the failing tests a maximum of `BUILDKITE_SPLITTER_RETRY_COUNT` times.  |
```

> | `BUILDKITE_SPLITTER_SPLIT_BY_EXAMPLE` | `false` | Flag to enable or disable split by example. When this option is `true`, the Test Splitter will split slow test files that take longer than 3 minutes into individual test cases, so that the execution of file will be over multiple nodes. |

Reviewer: ```suggestion
| `BUILDKITE_SPLITTER_SPLIT_BY_EXAMPLE` | `false` | Flag to enable split by example. When this option is `true`, the Test Splitter will split the execution of slow test files over multiple partitions. |
```

| `BUILDKITE_SPLITTER_TEST_CMD` | `bundle exec rspec {{testExamples}}` | Test command for running your tests. Test splitter will fill in the `{{testExamples}}` placeholder with the test splitting results |
| `BUILDKITE_SPLITTER_TEST_FILE_EXCLUDE_PATTERN` | - | Glob pattern to use for excluding test files or directory. </br> *It accepts pattern syntax supported by [zzglob](https://github.com/DrJosh9000/zzglob?tab=readme-ov-file#pattern-syntax) library.* |

> | `BUILDKITE_SPLITTER_TEST_FILE_PATTERN` | `spec/**/*_spec.rb` | Glob pattern for discovering test files that need to be executed. </br> *It accepts pattern syntax supported by [zzglob](https://github.com/DrJosh9000/zzglob?tab=readme-ov-file#pattern-syntax) library*. |

Reviewer: Which one of the include/exclude is applied first?

> chmod +x test-splitter

Reviewer: This bit can be deleted if we're recommending the user to run the test splitter directly via the step command

./test-splitter # fetches the test plan for this node, and then executes the rspec tests
```

To parallelise your tests in your Buildkite build, you can amend your pipeline step configuration to:
```
steps:
  - name: "Rspec"
    command: ./test-splitter
    parallelism: 10
    env:
      BUILDKITE_SPLITTER_SUITE_SLUG: my-suite

>       BUILDKITE_SPLITTER_API_ACCESS_TOKEN: your-secret-tokens

Reviewer: ```suggestion
      BUILDKITE_SPLITTER_API_ACCESS_TOKEN: your-secret-token
```
